### DISCUSSION TITLE
Jython 2.7 and the end of Python 2

### DISCUSSION THREAD

**Discussion by FUN-00000000** [NationalSecurityAgency/ghidra#3617] [General]
Hello, I was wondering if anyone has proposed any potential replacement for Jython 2.7.2.

I recently stumbled across [Issue #194](https://github.com/NationalSecurityAgency/ghidra/issues/194) which states that there is no official version of Jython 3 yet, and the newest version supported is still Jython 2.7.2. Additionally, another user recommended GraalPython, but it appears it may not be a suitable replacement either.

Overall, I think it would be useful to bring Python 3's changes to Ghidra, as Python 2 has reached end-of-life. If anyone has any ideas of how to do this, I'd love to hear them!

Thanks all

**Comment by bdemick** [comment]
I've done some very brief testing with [Ghidraal](https://github.com/jpleasu/ghidraal) and found it to be serviceable, although I think `PyPI` support is still a little limited with GraalVM. At this point, being tied to Python 2.7 is absolutely cumbersome, and I've considered just using the Python bindings for API exploration and prototyping, and then porting to Java. The major limitation there is that if you want to implement any kind of machine learning / data science, Java is a much, much different experience than Python.

**Comment by TorgoTorgo** [comment]
[Ghidra-Bridge](https://github.com/justfoxing/ghidra_bridge) provides python3 support via an RPC service, I use it a lot for prototyping and for integrating with other Python packages. The one thing to remember is if you're doing a large amount of iteration, use the `remote_eval` function to reduce the RPC requests.

I wrote a [Jupyter Notebook kernel](https://github.com/TorgoTorgo/ghidra-notebook) which is useful for prototyping, having Jupyter's result caching makes writing scripts and exploring the API really nice.

[Jython is adding python3 support](https://github.com/jython/jython/tree/main), so eventually I assume Ghidra will adopt this, so we won't be stuck with 2.7 forever.

Thanks for the Ghidraal link @bdemick, I'll give this a go. Sounds like a useful intermediate before Jython adds 3.8 support.

**Comment by fmagin** [comment]
If the concern is only about having a modern language for scripting tasks, I would throw in the plugin I wrote that embeds a Jupyter Kotlin kernel into Ghidra: https://github.com/GhidraJupyter/ghidra-jupyter-kotlin

The support for Kotlin in Jupyter in general is still fairly new, but under active development. IMO writing Kotlin isn't much harder than writing Python for most scripting tasks, and because Kotlin uses static types, features like tab completion work no matter how deeply nested some expression is. Unlike Python where this fails fairly quickly.

The downside is that this doesn't involve Python at all, so no chance of leveraging the existing Python ecosystem around datascience and machine learning libraries. Though at a glance, the libraries for Kotlin in this regard seem to cover the basics: https://kotlinlang.org/docs/data-science-overview.html#kotlin-libraries
